jax_md/elasticity.py

```diff
@@ -0,0 +1,586 @@
+  """ Setup calculation of elastic modulus tensor for a 2d or 3d athermal system.
+
+  The elastic modulus tensor describes a material's response to different 
+  boundary deformations. Specifically, for small deformation given by a 
+  symmetric strain tensor e, the change in energy is
+    U / V^0 = U^0/V^0 + s^0_{ij} e_{ji} + (1/2) C_{ijkl} e_{ij} e_{kl} + ...
+  where V^0 is the volume, U^0 is the initial energy, s^0 is the residual stress
+  tensor of the undeformed system, and C is the elastic modulus tensor. C is
+  a fourth-rank tensor of shape (dimension,dimension,dimension,dimension), with
+  the following symmetries.
+  
+    Minor symmetries:
+      C_ijkl = C_jikl = C_ijlk
+
+    Major symmetries:
+      C_ijkl = C_lkij
+
+  The minor symmetries are also reflected in the symmetric nature
+  of stress and strain tensors:
+    s_ij = s_ji
+    e_ij = e_ji
+
+  In general, there are 21 independent elastic constants in 3 dimension (6 in 2
+  dimensions). While systems with additional symmetries (e.g. isotropic,
+  orthotropic, etc.) can be expressed with fewer constants, we do not assume any
+  such additional symmetries.
+
+  At zero temperature, the response of every particle to a deformation can be 
+  calculated explicitly to linear order, enabling the exact calculation of 
+  the elastic modulus tensor without the need of any finite differences or any
+  approximations.
+
+  Args:
+    energy_fn: an energy function that is created using periodic_general so
+      that it can be called with and differentiated with respect to a 'box' 
+      argument. energy_fn(R, box=box) corresponds to the state around which we 
+      are expanding
+    tether_strength: scalar. Strength of the "tether" applied to each particle, 
+      which can be necessary to make the Hessian matrix non-singular. Solving
+      for the non-affine response of each particle requires that the Hessian is
+      positive definite. However, there can often be zero modes (eigenvectors of
+      the Hessian with zero eigenvalue) that do not couple to the boundary, and 
+      therefore do not affect the elastic constants despite the zero eigenvalue.
+      The most common example is the global translational modes. To solve for
+      the non-affine response, we consider the "tethered Hessian"
+        H + tether_strength * I, 
+      where I is the identity matrix and tether_strength is a small constant.
+    gradient_check: None or scalar. If not None, a check will be performed 
+      to guarantee that the maximum component of the gradient is less than
+      gradient_check. In other words, that
+        jnp.amax(jnp.abs(grad(energy_fn)(R, box=box))) < gradient_check == True
+      ***** Currently not implemented! *****
+    cg_tol: scalar. Tolorance used when solving for the non-affine response.
+  
+  Return: A function to calculate the elastic modulus tensor
+
+    
+  TODO:
+    - generalize to work with force_fn functions?
```

Comment:
Thinking about generalizing this to work with force functions rather than just energy functions... not sure it will work because line 241 needs to calculate the second derivative of the energy with respect to the strain magnitude `gamma`. I don't think you can back this out from a force function. Ideas or suggestions welcome!